### DISCUSSION TITLE
How to extend the exceptions error object?

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#13398] [Archive]
I tried to create a new class to inherit BaseException, but that didn't work, it still uses INTERNAL_SERVER_ERROR

## Not Working

<img width="824" alt="image" src="https://user-images.githubusercontent.com/101538735/169196774-3a0dc939-e289-4dfb-93c5-4480b99fc52e.png">

<img width="799" alt="image" src="https://user-images.githubusercontent.com/101538735/169194528-4a3f9a91-b08c-43d5-9b8a-2dc42fd0111d.png">

<img width="606" alt="image" src="https://user-images.githubusercontent.com/101538735/169194566-c20fd24c-9c9f-43ea-a443-c2f9276114f2.png">

## Working

I tried to shift to a new way of inheritance

<img width="1009" alt="image" src="https://user-images.githubusercontent.com/101538735/169195005-87ebbd85-3aee-4023-85f5-f0e7e14c1343.png">

<img width="505" alt="image" src="https://user-images.githubusercontent.com/101538735/169195104-28350d3c-19da-4c4e-9ef1-87b4ce2a7484.png">

it works! 
I don't understand why this is

**Comment by azrikahar** [comment]
I believe it's due to the `instanceof BaseException` in your first example will in fact be false, rather than true. When you shift the class declaration inside the export function, it then properly registers as an instance of the "same" BaseException. 

I'm not 100% sure on the reasoning myself, but it could somehow be related to https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work? cc @nickrum as I could be very wrong and way off here.

In your first example, if you were to use `require` instead of `import`, it _should_ work (at least it did for me):

```diff
- import { BaseException } from "@directus/shared/exceptions";
+ const { BaseException } = require("@directus/shared/exceptions");
```

**Reply by ghost** [reply]
yes you  are right "instanceof BaseException" always return false
 
i used "const {BaseException} = require("@directus/shared/exceptions");"

The BaseException exported from @directus/shared/exceptions is not the same as the BaseException inside the hooks

<img width="918" alt="image" src="https://user-images.githubusercontent.com/101538735/169244054-9dcce632-84e7-4171-b427-2140f979165e.png">


<img width="814" alt="image" src="https://user-images.githubusercontent.com/101538735/169244118-8e401df9-0d60-4129-89f1-db628f02b742.png">


<img width="836" alt="image" src="https://user-images.githubusercontent.com/101538735/169244138-af1dddcf-7f06-4793-8e13-4e1f328b813a.png">

**Reply by azrikahar** [reply]
Thanks for confirming! It is rather odd, and frankly I'm not 100% what's happening here either (aside from the link I shared earlier, as well as this link which seems to alude to compilation target being a potential culprit: https://www.dannyguo.com/blog/how-to-fix-instanceof-not-working-for-custom-errors-in-typescript/).

**Reply by nickrum** [reply]
Not exactly sure what's happening here either, but it could be that the extension and the API use different versions of the `@directus/shared` package.

**Comment by ghost** [comment]
I had to use the initExceptions function to save the BaseException returned inside the hook

<img width="820" alt="image" src="https://user-images.githubusercontent.com/101538735/169244601-d4b298f1-2cde-496d-bc22-99f8d55ed582.png">

<img width="485" alt="image" src="https://user-images.githubusercontent.com/101538735/169244629-3df9c433-c5fb-4b7e-958d-b267abb45c9d.png">

<img width="618" alt="image" src="https://user-images.githubusercontent.com/101538735/169244661-40f618de-20f7-4f00-89f5-a8c9f29d592e.png">
